File: frontend/src/Components/Requests/ViewAllScrapersRequest.js
import React, { useEffect, useState } from "react";
import server from "../Static/Constants";

const ViewAllScrapersRequest = () => {
  const [scrapers, setScrapers] = useState([]);
  const [errorMessage, setErrorMessage] = useState(null);
  const [successMessage, setSuccessMessage] = useState(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    const fetchScrapers = async () => {
      try {
        const response = await fetch(`${server}/get/all_scraper_statuses/`, {
          method: "GET",
          headers: {
            Authorization: localStorage.getItem("token"),
          },
        });

        if (!response.ok) {
          throw new Error("Failed to fetch scrapers");
        }

        const data = await response.json();

        const scraperArray = Object.entries(data.scrapers || {}).map(
          ([serial, info]) => ({
            serial,
            ...info,
          })
        );

        setScrapers(scraperArray);
      } catch (error) {
        console.error("Error fetching scrapers:", error);
        setErrorMessage("Error fetching scrapers");
      }
    };

    fetchScrapers();
  }, []);

  const handleRunScraper = async (serial) => {
    setErrorMessage(null);
    setSuccessMessage(null);
    setLoading(true);

    try {
      const response = await fetch(`${server}/trigger/mtg_f2f_scraper/${serial}/`, {
        method: "POST",
        headers: {
          Authorization: localStorage.getItem("token"),
        },
      });

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.error || "Failed to trigger scraper");
      }

      setSuccessMessage(`Scraper ${serial} triggered successfully.`);
    } catch (error) {
      console.error("Error triggering scraper:", error);
      setErrorMessage(error.message);
    } finally {
      setLoading(false);
    }
  };

  return (
    <div>
      <h1>All Scrapers</h1>

      {errorMessage && <p style={{ color: "red" }}>{errorMessage}</p>}
      {successMessage && <p style={{ color: "green" }}>{successMessage}</p>}

      {scrapers.length === 0 ? (
        <p>No scrapers found.</p>
      ) : (
        <ul>
          {scrapers.map((scraper) => (
            <li key={scraper.serial} style={{ marginBottom: "1em" }}>
              <strong>{scraper.file_name}</strong> — {scraper.status} <br />
              <small>
                Created: {new Date(scraper.create_date).toLocaleString()}
              </small>
              <br />

              <button
                onClick={() => handleRunScraper(scraper.serial)}
                disabled={loading || scraper.status !== "validated"}
                style={{
                  marginTop: "0.5em",
                  marginRight: "0.5em",
                  padding: "0.4em 0.8em",
                  backgroundColor:
                    scraper.status === "validated" ? "#4CAF50" : "#aaa",
                  color: "white",
                  border: "none",
                  borderRadius: "4px",
                  cursor:
                    scraper.status === "validated" && !loading
                      ? "pointer"
                      : "not-allowed",
                }}
              >
                {loading ? "Processing..." : "Run Scraper"}
              </button>

              <a
                href={`/mtg/viewscraperhistory/?serial=${scraper.serial}`}
                style={{
                  display: "inline-block",
                  padding: "0.4em 0.8em",
                  backgroundColor: "#2196F3",
                  color: "white",
                  borderRadius: "4px",
                  textDecoration: "none",
                  cursor: "pointer",
                }}
              >
                View Previous Sessions
              </a>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default ViewAllScrapersRequest;

File: flutter_frontend/lib/assets/mtg/requests/viewAllScrapersRequest.dart
import 'package:flutter/material.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_frontend/static/constants.dart';
import 'package:flutter_frontend/assets/mtg/requests/viewScraperHistoryRequest.dart';

class ViewAllScraperRequest extends StatefulWidget {
  const ViewAllScraperRequest({Key? key}) : super(key: key);

  @override
  State<ViewAllScraperRequest> createState() => _ViewAllScraperRequestState();
}

class _ViewAllScraperRequestState extends State<ViewAllScraperRequest> {
  final _storage = const FlutterSecureStorage();
  List<dynamic> _scrapers = [];
  String? _errorMessage;
  String? _successMessage;
  bool _loading = false;
  bool _triggering = false;

  @override
  void initState() {
    super.initState();
    _fetchScrapers();
  }

  Future<void> _fetchScrapers() async {
    setState(() {
      _loading = true;
      _errorMessage = null;
      _successMessage = null;
    });

    try {
      final token = await _storage.read(key: 'token');
      final response = await http.get(
        Uri.parse('$server/get/all_scraper_statuses/'),
        headers: {'Authorization': token ?? ''},
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final scrapersMap = data['scrapers'] as Map<String, dynamic>? ?? {};
        setState(() {
          _scrapers = scrapersMap.entries
              .map((e) => {'serial': e.key, ...e.value})
              .toList();
        });
      } else {
        setState(() {
          _errorMessage = 'Failed to fetch scrapers: ${response.reasonPhrase}';
        });
      }
    } catch (e) {
      setState(() {
        _errorMessage = 'Error fetching scrapers: $e';
      });
    } finally {
      setState(() {
        _loading = false;
      });
    }
  }

  Future<void> _runScraper(String serial) async {
    setState(() {
      _errorMessage = null;
      _successMessage = null;
      _triggering = true;
    });

    try {
      final token = await _storage.read(key: 'token');
      final response = await http.post(
        Uri.parse('$server/trigger/mtg_f2f_scraper/$serial/'),
        headers: {'Authorization': token ?? ''},
      );

      final data = jsonDecode(response.body);

      if (response.statusCode == 200) {
        setState(() {
          _successMessage = 'Scraper $serial triggered successfully.';
        });
      } else {
        setState(() {
          _errorMessage = data['error'] ?? 'Failed to trigger scraper.';
        });
      }
    } catch (e) {
      setState(() {
        _errorMessage = 'Error triggering scraper: $e';
      });
    } finally {
      setState(() {
        _triggering = false;
      });
    }
  }

  void _viewPreviousSessions(String serial) async {
    try {
      await _storage.write(key: 'scraperSerial', value: serial);

      Navigator.push(
        context,
        MaterialPageRoute(
          builder: (context) => const ViewScraperHistoryRequest(),
        ),
      );
    } catch (e) {
      setState(() {
        _errorMessage = 'Error opening scraper history: $e';
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('All Scrapers')),
      body: Padding(
        padding: const EdgeInsets.all(16),
        child: _loading
            ? const Center(child: CircularProgressIndicator())
            : Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: [
                  if (_errorMessage != null)
                    Padding(
                      padding: const EdgeInsets.only(bottom: 8),
                      child: Text(
                        _errorMessage!,
                        style: const TextStyle(color: Colors.red),
                        textAlign: TextAlign.center,
                      ),
                    ),
                  if (_successMessage != null)
                    Padding(
                      padding: const EdgeInsets.only(bottom: 8),
                      child: Text(
                        _successMessage!,
                        style: const TextStyle(color: Colors.green),
                        textAlign: TextAlign.center,
                      ),
                    ),
                  Expanded(
                    child: _scrapers.isEmpty
                        ? const Center(child: Text('No scrapers found.'))
                        : ListView.builder(
                            itemCount: _scrapers.length,
                            itemBuilder: (context, index) {
                              final scraper = _scrapers[index];
                              final status = scraper['status'] ?? '';
                              final serial = scraper['serial'] ?? '';
                              final fileName = scraper['file_name'] ?? '';
                              final createDate = scraper['create_date'] ?? '';

                              return Card(
                                margin: const EdgeInsets.symmetric(vertical: 8),
                                child: Padding(
                                  padding: const EdgeInsets.all(12),
                                  child: Column(
                                    crossAxisAlignment:
                                        CrossAxisAlignment.start,
                                    children: [
                                      Text(
                                        fileName,
                                        style: const TextStyle(
                                          fontWeight: FontWeight.bold,
                                        ),
                                      ),
                                      const SizedBox(height: 4),
                                      Text('Status: $status'),
                                      const SizedBox(height: 4),
                                      Text(
                                        'Created: ${DateTime.tryParse(createDate)?.toLocal().toString() ?? createDate}',
                                      ),
                                      const SizedBox(height: 8),
                                      Row(
                                        children: [
                                          ElevatedButton(
                                            onPressed:
                                                status == 'validated' &&
                                                    !_triggering
                                                ? () => _runScraper(serial)
                                                : null,
                                            style: ElevatedButton.styleFrom(
                                              backgroundColor:
                                                  status == 'validated'
                                                  ? Colors.green
                                                  : Colors.grey,
                                            ),
                                            child: _triggering
                                                ? const SizedBox(
                                                    width: 16,
                                                    height: 16,
                                                    child:
                                                        CircularProgressIndicator(
                                                          strokeWidth: 2,
                                                          color: Colors.white,
                                                        ),
                                                  )
                                                : const Text('Run Scraper'),
                                          ),
                                          const SizedBox(width: 8),
                                          TextButton(
                                            onPressed: () =>
                                                _viewPreviousSessions(serial),
                                            style: TextButton.styleFrom(
                                              backgroundColor: Colors.blue,
                                              foregroundColor: Colors.white,
                                            ),
                                            child: const Text(
                                              'View Previous Sessions',
                                            ),
                                          ),
                                        ],
                                      ),
                                    ],
                                  ),
                                ),
                              );
                            },
                          ),
                  ),
                ],
              ),
      ),
    );
  }
}
